linux/src/threadpool_linux.c

```diff
@@ -614,7 +616,49 @@ int threadpool_schedule_work(THANDLE(THREADPOOL) threadpool, THREADPOOL_WORK_FUN
     return result;
 }
 
-int threadpool_timer_start(THANDLE(THREADPOOL) threadpool, uint32_t start_delay_ms, uint32_t timer_period_ms, THREADPOOL_WORK_FUNCTION work_function, void* work_function_ctx, TIMER_INSTANCE_HANDLE* timer_handle)
+static void threadpool_timer_dispose(TIMER_INSTANCE_HANDLE timer)
+{
+    while (true)
+    {
+        /* Codes_SRS_THREADPOOL_LINUX_45_008: [ threadpool_timer_destroy shall call InterlockedHL_WaitForNotValue to wait until timer_work_guard is not TIMER_WORKING. ]*/
+        INTERLOCKED_HL_RESULT guard_result = InterlockedHL_WaitForNotValue(&timer->timer_work_guard, TIMER_WORKING, UINT32_MAX);
+        if (guard_result == INTERLOCKED_HL_OK)
+        {
+            /* Codes_SRS_THREADPOOL_LINUX_45_009: [ threadpool_timer_destroy shall call interlocked_add to add 0 to timer_work_guard to get current value of timer_work_guard. ]*/
+            TIMER_GUARD guard_value = interlocked_add(&timer->timer_work_guard, 0);
+            if (guard_value != TIMER_WORKING)
+            {
+                /* Codes_SRS_THREADPOOL_LINUX_45_010: [ threadpool_timer_destroy shall call interlocked_compare_exchange on timer_work_guard with the current value of timer_work_guard as the comparison and TIMER_DELETING as the exchange. ]*/
+                if (interlocked_compare_exchange(&timer->timer_work_guard, TIMER_DELETING, guard_value) == guard_value)
+                {
+                    /* Codes_SRS_THREADPOOL_LINUX_45_007: [ Until timer_work_guard can be set to TIMER_DELETING. ]*/
+                    break;
+                }
+            }
+        }
+    }
+    /* Codes_SRS_THREADPOOL_LINUX_07_071: [ threadpool_timer_cancel shall call timer_delete to destroy the ongoing timers. ]*/
+    if (timer_delete(timer->time_id) != 0)
+    {
+        LogErrorNo("Failure calling timer_delete.");
+    }
+    else
+    {
+        // Do Nothing
+    }
+    // Even though timer_delete does cancel any events, there is a small window where an event was triggered just before and a thread is being created.
+    // So the callback can still execute after the timer_delete call. A small wait here keeps the timer instance alive long enough for
+    // the callback to use it if we hit this window.
+    /* Codes_SRS_THREADPOOL_LINUX_45_012: [ threadpool_timer_cancel shall call ThreadAPI_Sleep to allow timer resources to clean up. ]*/
+    ThreadAPI_Sleep(10);
+    // This doesn't fix the problem, this won't 100% guarantee the thread will be executed before we free the timer data. There are a couple of
+    // options to fix the problem, but the way that fits this threadpool model best would be to keep a pool of timers in the threadpool,
+    // and manage each state as in-use or not. That way the timer data allocation is not dependent on threadpool_timer_start/destroy.
+    // To reproduce this problem, create a large number of threads that delete the timer at the same time it is due to expire and remove the pause
+    // above.
```

Comment:
It looks like we're making fixes where we use a lock to ensure safe access to state only on Windows... Why not on Linux also?

This part specifically with the sleep is not very cool.

Consider a separate task for it, but I think the same "fix" should be applied in Linux.

Reply:
added the fix to linux